src/librustc/ty/print/pretty.rs

```diff
@@ -969,69 +965,138 @@ pub trait PrettyPrinter<'tcx>:
                     _ => p!(write("{}{}", sign_extend(data, size) as i128, i_str))
                 }
             },
-            (ConstValue::Scalar(Scalar::Raw { data, .. }), ty::Char) =>
+            (Scalar::Raw { data, .. }, ty::Char) =>
                 p!(write("{:?}", ::std::char::from_u32(data as u32).unwrap())),
-            (ConstValue::Scalar(_), ty::RawPtr(_)) => p!(write("{{pointer}}")),
-            (ConstValue::Scalar(Scalar::Ptr(ptr)), ty::FnPtr(_)) => {
+            (Scalar::Raw { data: 0, .. }, ty::RawPtr(_)) => p!(write("{{null pointer}}")),
+            (Scalar::Raw { data, .. }, ty::RawPtr(_)) => {
+                let pointer_width = self.tcx().data_layout.pointer_size.bytes();
+                p!(write("0x{:01$x}", data, pointer_width as usize))
+            },
+            (Scalar::Ptr(p), ty::RawPtr(_)) => self = self.pretty_print_const_pointer(p)?,
```

Comment:
So for raw ptrs we cover all the cases here, but strangely not for references?

Reply:
I assume that for references we'll want to print the content? I wasn't sure what we want to do so I kept the status quo: debug printing.

Reply:
So raw ptrs now honor the "show alloc id" flag but references always debug print?

Reply:
yes

Reply:
That's rather inconsistent, isn't it?

Reply:
well... for MIR it doesn't matter, and for diagnostics I don't think we actually end up with these anywhere other than `&str` and `&[u8]`

Reply:
Any reason *not* to reorder the branches here a bit to handle all reference-special-cases first, and then use `ty::RawPtr | ty::Ref` to cover all cases?

Reply:
done